// @deepgram/.ignored_sdk/dist/esm/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AgentClient } from "./api/resources/agent/client/Client.mjs";
import { AuthClient } from "./api/resources/auth/client/Client.mjs";
import { ListenClient } from "./api/resources/listen/client/Client.mjs";
import { ManageClient } from "./api/resources/manage/client/Client.mjs";
import { ReadClient } from "./api/resources/read/client/Client.mjs";
import { SelfHostedClient } from "./api/resources/selfHosted/client/Client.mjs";
import { SpeakClient } from "./api/resources/speak/client/Client.mjs";
import { VoiceAgentClient } from "./api/resources/voiceAgent/client/Client.mjs";
import { normalizeClientOptionsWithAuth } from "./BaseClient.mjs";
import * as core from "./core/index.mjs";
import * as environments from "./environments.mjs";
export class DeepgramClient {
    constructor(options = {}) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new AgentClient(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new AuthClient(this._options)));
    }
    get listen() {
        var _a;
        return ((_a = this._listen) !== null && _a !== void 0 ? _a : (this._listen = new ListenClient(this._options)));
    }
    get manage() {
        var _a;
        return ((_a = this._manage) !== null && _a !== void 0 ? _a : (this._manage = new ManageClient(this._options)));
    }
    get read() {
        var _a;
        return ((_a = this._read) !== null && _a !== void 0 ? _a : (this._read = new ReadClient(this._options)));
    }
    get selfHosted() {
        var _a;
        return ((_a = this._selfHosted) !== null && _a !== void 0 ? _a : (this._selfHosted = new SelfHostedClient(this._options)));
    }
    get speak() {
        var _a;
        return ((_a = this._speak) !== null && _a !== void 0 ? _a : (this._speak = new SpeakClient(this._options)));
    }
    get voiceAgent() {
        var _a;
        return ((_a = this._voiceAgent) !== null && _a !== void 0 ? _a : (this._voiceAgent = new VoiceAgentClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                // Patch: the 2026-05-14 regen switched this default from `base`
                // (api.deepgram.com) to `agentRest` (agent.deepgram.com) after
                // the new agentRest slot was introduced upstream. The passthrough
                // helper is documented as the catch-all for endpoints not yet
                // supported in the SDK — it should default to the canonical
                // Deepgram REST host, not the agent host. Restoring `base`.
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : (() => __awaiter(this, void 0, void 0, function* () {
                    var _a;
                    const env = yield core.Supplier.get(this._options.environment);
                    return typeof env === "string"
                        ? env
                        : ((_a = env === null || env === void 0 ? void 0 : env.base) !== null && _a !== void 0 ? _a : environments.DeepgramEnvironment.Production.base);
                })),
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
