// @elevenlabs/.ignored_elevenlabs-js/dist/wrapper/speech-engine/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeechEngineClientWrapper = exports.SpeechEngineAttachment = exports.SpeechEngineResource = exports.SpeechEngineServer = exports.SpeechEngineSession = exports.SpeechEngine = void 0;
exports.SpeechEngine = __importStar(require("./namespace"));
var SpeechEngineSession_1 = require("./SpeechEngineSession");
Object.defineProperty(exports, "SpeechEngineSession", { enumerable: true, get: function () { return SpeechEngineSession_1.SpeechEngineSession; } });
var SpeechEngineServer_1 = require("./SpeechEngineServer");
Object.defineProperty(exports, "SpeechEngineServer", { enumerable: true, get: function () { return SpeechEngineServer_1.SpeechEngineServer; } });
var SpeechEngineResource_1 = require("./SpeechEngineResource");
Object.defineProperty(exports, "SpeechEngineResource", { enumerable: true, get: function () { return SpeechEngineResource_1.SpeechEngineResource; } });
var SpeechEngineAttachment_1 = require("./SpeechEngineAttachment");
Object.defineProperty(exports, "SpeechEngineAttachment", { enumerable: true, get: function () { return SpeechEngineAttachment_1.SpeechEngineAttachment; } });
var SpeechEngineClientWrapper_1 = require("./SpeechEngineClientWrapper");
Object.defineProperty(exports, "SpeechEngineClientWrapper", { enumerable: true, get: function () { return SpeechEngineClientWrapper_1.SpeechEngineClientWrapper; } });
