// @deepgram/.ignored_sdk/dist/esm/api/resources/listen/resources/v2/client/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { normalizeClientOptions } from "../../../../../../BaseClient.mjs";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../../../core/headers.mjs";
import * as core from "../../../../../../core/index.mjs";
import { toJson } from "../../../../../../core/json.mjs";
import * as environments from "../../../../../../environments.mjs";
import { V2Socket } from "./Socket.mjs";
export class V2Client {
    constructor(options = {}) {
        this._options = normalizeClientOptions(options);
    }
    connect(args) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c;
            const { model, encoding, sample_rate: sampleRate, eager_eot_threshold: eagerEotThreshold, eot_threshold: eotThreshold, eot_timeout_ms: eotTimeoutMs, keyterm, language_hint: languageHint, profanity_filter: profanityFilter, mip_opt_out: mipOptOut, tag, protocols, queryParams, headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal, } = args;
            const _queryParams = {
                model,
                encoding: encoding != null ? encoding : undefined,
                sample_rate: sampleRate != null ? (typeof sampleRate === "string" ? sampleRate : toJson(sampleRate)) : undefined,
                eager_eot_threshold: eagerEotThreshold != null
                    ? typeof eagerEotThreshold === "string"
                        ? eagerEotThreshold
                        : toJson(eagerEotThreshold)
                    : undefined,
                eot_threshold: eotThreshold != null
                    ? typeof eotThreshold === "string"
                        ? eotThreshold
                        : toJson(eotThreshold)
                    : undefined,
                eot_timeout_ms: eotTimeoutMs != null
                    ? typeof eotTimeoutMs === "string"
                        ? eotTimeoutMs
                        : toJson(eotTimeoutMs)
                    : undefined,
                keyterm: Array.isArray(keyterm)
                    ? keyterm.map((item) => (typeof item === "string" ? item : toJson(item)))
                    : keyterm != null
                        ? typeof keyterm === "string"
                            ? keyterm
                            : toJson(keyterm)
                        : undefined,
                language_hint: Array.isArray(languageHint)
                    ? languageHint.map((item) => (typeof item === "string" ? item : toJson(item)))
                    : languageHint != null
                        ? typeof languageHint === "string"
                            ? languageHint
                            : toJson(languageHint)
                        : undefined,
                profanity_filter: profanityFilter != null ? profanityFilter : undefined,
                mip_opt_out: mipOptOut != null ? (typeof mipOptOut === "string" ? mipOptOut : toJson(mipOptOut)) : undefined,
                tag: tag != null ? (typeof tag === "string" ? tag : toJson(tag)) : undefined,
            };
            const _headers = mergeHeaders((_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, mergeOnlyDefinedHeaders({ Authorization: args.Authorization }), headers);
            const socket = new core.ReconnectingWebSocket({
                url: core.url.join((_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : ((_c = (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.DeepgramEnvironment.Production).production, "/v2/listen"),
                protocols: protocols !== null && protocols !== void 0 ? protocols : [],
                queryParameters: Object.assign(Object.assign({}, _queryParams), queryParams),
                headers: _headers,
                options: {
                    debug: debug !== null && debug !== void 0 ? debug : false,
                    maxRetries: reconnectAttempts !== null && reconnectAttempts !== void 0 ? reconnectAttempts : 30,
                    connectionTimeout: connectionTimeoutInSeconds != null ? connectionTimeoutInSeconds * 1000 : undefined,
                },
                abortSignal: abortSignal,
            });
            return new V2Socket({ socket });
        });
    }
}
