// @deepgram/.ignored_sdk/dist/cjs/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeepgramClient = void 0;
const Client_js_1 = require("./api/resources/agent/client/Client.js");
const Client_js_2 = require("./api/resources/auth/client/Client.js");
const Client_js_3 = require("./api/resources/listen/client/Client.js");
const Client_js_4 = require("./api/resources/manage/client/Client.js");
const Client_js_5 = require("./api/resources/read/client/Client.js");
const Client_js_6 = require("./api/resources/selfHosted/client/Client.js");
const Client_js_7 = require("./api/resources/speak/client/Client.js");
const Client_js_8 = require("./api/resources/voiceAgent/client/Client.js");
const BaseClient_js_1 = require("./BaseClient.js");
const core = __importStar(require("./core/index.js"));
const environments = __importStar(require("./environments.js"));
class DeepgramClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new Client_js_1.AgentClient(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new Client_js_2.AuthClient(this._options)));
    }
    get listen() {
        var _a;
        return ((_a = this._listen) !== null && _a !== void 0 ? _a : (this._listen = new Client_js_3.ListenClient(this._options)));
    }
    get manage() {
        var _a;
        return ((_a = this._manage) !== null && _a !== void 0 ? _a : (this._manage = new Client_js_4.ManageClient(this._options)));
    }
    get read() {
        var _a;
        return ((_a = this._read) !== null && _a !== void 0 ? _a : (this._read = new Client_js_5.ReadClient(this._options)));
    }
    get selfHosted() {
        var _a;
        return ((_a = this._selfHosted) !== null && _a !== void 0 ? _a : (this._selfHosted = new Client_js_6.SelfHostedClient(this._options)));
    }
    get speak() {
        var _a;
        return ((_a = this._speak) !== null && _a !== void 0 ? _a : (this._speak = new Client_js_7.SpeakClient(this._options)));
    }
    get voiceAgent() {
        var _a;
        return ((_a = this._voiceAgent) !== null && _a !== void 0 ? _a : (this._voiceAgent = new Client_js_8.VoiceAgentClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                // Patch: the 2026-05-14 regen switched this default from `base`
                // (api.deepgram.com) to `agentRest` (agent.deepgram.com) after
                // the new agentRest slot was introduced upstream. The passthrough
                // helper is documented as the catch-all for endpoints not yet
                // supported in the SDK — it should default to the canonical
                // Deepgram REST host, not the agent host. Restoring `base`.
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : (() => __awaiter(this, void 0, void 0, function* () {
                    var _a;
                    const env = yield core.Supplier.get(this._options.environment);
                    return typeof env === "string"
                        ? env
                        : ((_a = env === null || env === void 0 ? void 0 : env.base) !== null && _a !== void 0 ? _a : environments.DeepgramEnvironment.Production.base);
                })),
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
exports.DeepgramClient = DeepgramClient;
