// @elevenlabs/.ignored_elevenlabs-js/dist/serialization/resources/workspace/resources/authConnections/types/AuthConnectionsCreateRequestBody.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthConnectionsCreateRequestBody = void 0;
const core = __importStar(require("../../../../../../core"));
const CreateBasicAuthRequest_1 = require("../../../../../types/CreateBasicAuthRequest");
const CreateBearerAuthRequest_1 = require("../../../../../types/CreateBearerAuthRequest");
const CreateCustomHeaderAuthRequest_1 = require("../../../../../types/CreateCustomHeaderAuthRequest");
const CreateMtlsAuthRequest_1 = require("../../../../../types/CreateMtlsAuthRequest");
const CreateOAuth2ClientCredsRequest_1 = require("../../../../../types/CreateOAuth2ClientCredsRequest");
const CreateOAuth2JwtRequest_1 = require("../../../../../types/CreateOAuth2JwtRequest");
const CreatePrivateKeyJwtRequest_1 = require("../../../../../types/CreatePrivateKeyJwtRequest");
exports.AuthConnectionsCreateRequestBody = core.serialization
    .union(core.serialization.discriminant("authType", "auth_type"), {
    oauth2_client_credentials: CreateOAuth2ClientCredsRequest_1.CreateOAuth2ClientCredsRequest,
    custom_header_auth: CreateCustomHeaderAuthRequest_1.CreateCustomHeaderAuthRequest,
    basic_auth: CreateBasicAuthRequest_1.CreateBasicAuthRequest,
    bearer_auth: CreateBearerAuthRequest_1.CreateBearerAuthRequest,
    oauth2_jwt: CreateOAuth2JwtRequest_1.CreateOAuth2JwtRequest,
    private_key_jwt: CreatePrivateKeyJwtRequest_1.CreatePrivateKeyJwtRequest,
    mtls: CreateMtlsAuthRequest_1.CreateMtlsAuthRequest,
})
    .transform({
    transform: (value) => value,
    untransform: (value) => value,
});
