// @elevenlabs/.ignored_elevenlabs-js/dist/serialization/resources/speechEngine/client/requests/CreateSpeechEngineRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateSpeechEngineRequest = void 0;
const core = __importStar(require("../../../../../core"));
const AgentCallLimits_1 = require("../../../../types/AgentCallLimits");
const AsrConversationalConfig_1 = require("../../../../types/AsrConversationalConfig");
const BaseTurnConfig_1 = require("../../../../types/BaseTurnConfig");
const ConversationConfigInput_1 = require("../../../../types/ConversationConfigInput");
const PrivacyConfigInput_1 = require("../../../../types/PrivacyConfigInput");
const SpeechEngineConfig_1 = require("../../../../types/SpeechEngineConfig");
const SpeechEngineConversationInitiationClientDataConfig_1 = require("../../../../types/SpeechEngineConversationInitiationClientDataConfig");
const TtsConversationalConfigInput_1 = require("../../../../types/TtsConversationalConfigInput");
exports.CreateSpeechEngineRequest = core.serialization.object({
    name: core.serialization.string().optional(),
    speechEngine: core.serialization.property("speech_engine", SpeechEngineConfig_1.SpeechEngineConfig),
    asr: AsrConversationalConfig_1.AsrConversationalConfig.optional(),
    tts: TtsConversationalConfigInput_1.TtsConversationalConfigInput.optional(),
    turn: BaseTurnConfig_1.BaseTurnConfig.optional(),
    conversation: ConversationConfigInput_1.ConversationConfigInput.optional(),
    privacy: PrivacyConfigInput_1.PrivacyConfigInput.optional(),
    callLimits: core.serialization.property("call_limits", AgentCallLimits_1.AgentCallLimits.optional()),
    language: core.serialization.string().optional(),
    tags: core.serialization.list(core.serialization.string()).optional(),
    overrides: SpeechEngineConversationInitiationClientDataConfig_1.SpeechEngineConversationInitiationClientDataConfig.optional(),
});
