// @elevenlabs/.ignored_elevenlabs-js/dist/api/resources/dubbing/client/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DubbingClient = void 0;
const BaseClient_1 = require("../../../../BaseClient");
const core = __importStar(require("../../../../core"));
const headers_1 = require("../../../../core/headers");
const environments = __importStar(require("../../../../environments"));
const handleNonStatusCodeError_1 = require("../../../../errors/handleNonStatusCodeError");
const errors = __importStar(require("../../../../errors/index"));
const serializers = __importStar(require("../../../../serialization/index"));
const ElevenLabs = __importStar(require("../../../index"));
const Client_1 = require("../resources/audio/client/Client");
const Client_2 = require("../resources/resource/client/Client");
const Client_3 = require("../resources/transcript/client/Client");
const Client_4 = require("../resources/transcripts/client/Client");
class DubbingClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_1.normalizeClientOptions)(options);
    }
    get resource() {
        var _a;
        return ((_a = this._resource) !== null && _a !== void 0 ? _a : (this._resource = new Client_2.ResourceClient(this._options)));
    }
    get audio() {
        var _a;
        return ((_a = this._audio) !== null && _a !== void 0 ? _a : (this._audio = new Client_1.AudioClient(this._options)));
    }
    get transcript() {
        var _a;
        return ((_a = this._transcript) !== null && _a !== void 0 ? _a : (this._transcript = new Client_3.TranscriptClient(this._options)));
    }
    get transcripts() {
        var _a;
        return ((_a = this._transcripts) !== null && _a !== void 0 ? _a : (this._transcripts = new Client_4.TranscriptsClient(this._options)));
    }
    /**
     * List the dubs you have access to.
     *
     * @param {ElevenLabs.DubbingListRequest} request
     * @param {DubbingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link ElevenLabs.UnprocessableEntityError}
     *
     * @example
     *     await client.dubbing.list({
     *         cursor: "cursor",
     *         pageSize: 1,
     *         dubbingStatus: "dubbing",
     *         dubbingStatuses: ["queued"],
     *         dubbingModels: ["dubbing_v1"],
     *         targetLanguageCodes: ["target_language_codes"],
     *         creationSources: ["flow_node"],
     *         filterByCreator: "personal",
     *         orderBy: "created_at",
     *         orderDirection: "DESCENDING"
     *     })
     */
    list(request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__list(request, requestOptions));
    }
    __list() {
        return __awaiter(this, arguments, void 0, function* (request = {}, requestOptions) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
            const { cursor, pageSize, dubbingStatus, dubbingStatuses, dubbingModels, targetLanguageCodes, creationSources, filterByCreator, orderBy, orderDirection, } = request;
            const _queryParams = {
                cursor,
                page_size: pageSize,
                dubbing_status: dubbingStatus != null
                    ? serializers.DubbingListRequestDubbingStatus.jsonOrThrow(dubbingStatus, {
                        unrecognizedObjectKeys: "strip",
                    })
                    : undefined,
                dubbing_statuses: Array.isArray(dubbingStatuses)
                    ? dubbingStatuses.map((item) => serializers.DubbingListRequestDubbingStatusesItem.jsonOrThrow(item, {
                        unrecognizedObjectKeys: "strip",
                    }))
                    : dubbingStatuses != null
                        ? serializers.DubbingListRequestDubbingStatusesItem.jsonOrThrow(dubbingStatuses, {
                            unrecognizedObjectKeys: "strip",
                        })
                        : undefined,
                dubbing_models: Array.isArray(dubbingModels)
                    ? dubbingModels.map((item) => serializers.DubbingListRequestDubbingModelsItem.jsonOrThrow(item, {
                        unrecognizedObjectKeys: "strip",
                    }))
                    : dubbingModels != null
                        ? serializers.DubbingListRequestDubbingModelsItem.jsonOrThrow(dubbingModels, {
                            unrecognizedObjectKeys: "strip",
                        })
                        : undefined,
                target_language_codes: targetLanguageCodes,
                creation_sources: Array.isArray(creationSources)
                    ? creationSources.map((item) => serializers.DubbingListRequestCreationSourcesItem.jsonOrThrow(item, {
                        unrecognizedObjectKeys: "strip",
                    }))
                    : creationSources != null
                        ? serializers.DubbingListRequestCreationSourcesItem.jsonOrThrow(creationSources, {
                            unrecognizedObjectKeys: "strip",
                        })
                        : undefined,
                filter_by_creator: filterByCreator != null
                    ? serializers.DubbingListRequestFilterByCreator.jsonOrThrow(filterByCreator, {
                        unrecognizedObjectKeys: "strip",
                    })
                    : undefined,
                order_by: orderBy != null
                    ? serializers.DubbingListRequestOrderBy.jsonOrThrow(orderBy, { unrecognizedObjectKeys: "strip" })
                    : undefined,
                order_direction: orderDirection != null
                    ? serializers.DubbingListRequestOrderDirection.jsonOrThrow(orderDirection, {
                        unrecognizedObjectKeys: "strip",
                    })
                    : undefined,
            };
            const _headers = (0, headers_1.mergeHeaders)((_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "xi-api-key": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiKey) !== null && _b !== void 0 ? _b : (_c = this._options) === null || _c === void 0 ? void 0 : _c.apiKey }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_d = this._options.fetcher) !== null && _d !== void 0 ? _d : core.fetcher)({
                url: core.url.join((_f = (_e = (yield core.Supplier.get(this._options.baseUrl))) !== null && _e !== void 0 ? _e : (yield core.Supplier.get(this._options.environment))) !== null && _f !== void 0 ? _f : environments.ElevenLabsEnvironment.Production, "v1/dubbing"),
                method: "GET",
                headers: _headers,
                queryParameters: Object.assign(Object.assign({}, _queryParams), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams),
                timeoutMs: ((_j = (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.timeoutInSeconds) !== null && _j !== void 0 ? _j : 240) * 1000,
                maxRetries: (_k = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _k !== void 0 ? _k : (_l = this._options) === null || _l === void 0 ? void 0 : _l.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_m = this._options) === null || _m === void 0 ? void 0 : _m.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.DubbingMetadataPageResponseModel.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 422:
                        throw new ElevenLabs.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.ElevenLabsError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "GET", "/v1/dubbing");
        });
    }
    /**
     * Dubs a provided audio or video file into given language.
     *
     * @param {ElevenLabs.BodyDubAVideoOrAnAudioFileV1DubbingPost} request
     * @param {DubbingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link ElevenLabs.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.dubbing.create({})
     */
    create(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__create(request, requestOptions));
    }
    __create(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x;
            const _body = yield core.newFormData();
            if (request.file != null) {
                yield _body.appendFile("file", request.file);
            }
            if (request.csvFile != null) {
                yield _body.appendFile("csv_file", request.csvFile);
            }
            if (request.foregroundAudioFile != null) {
                yield _body.appendFile("foreground_audio_file", request.foregroundAudioFile);
            }
            if (request.backgroundAudioFile != null) {
                yield _body.appendFile("background_audio_file", request.backgroundAudioFile);
            }
            if (request.name != null) {
                _body.append("name", request.name);
            }
            if (request.sourceUrl != null) {
                _body.append("source_url", request.sourceUrl);
            }
            if (request.sourceLang != null) {
                _body.append("source_lang", request.sourceLang);
            }
            if (request.targetLang != null) {
                _body.append("target_lang", request.targetLang);
            }
            if (request.targetAccent != null) {
                _body.append("target_accent", request.targetAccent);
            }
            if (request.numSpeakers != null) {
                _body.append("num_speakers", (_a = request.numSpeakers) === null || _a === void 0 ? void 0 : _a.toString());
            }
            if (request.watermark != null) {
                _body.append("watermark", (_b = request.watermark) === null || _b === void 0 ? void 0 : _b.toString());
            }
            if (request.startTime != null) {
                _body.append("start_time", (_c = request.startTime) === null || _c === void 0 ? void 0 : _c.toString());
            }
            if (request.endTime != null) {
                _body.append("end_time", (_d = request.endTime) === null || _d === void 0 ? void 0 : _d.toString());
            }
            if (request.highestResolution != null) {
                _body.append("highest_resolution", (_e = request.highestResolution) === null || _e === void 0 ? void 0 : _e.toString());
            }
            if (request.dropBackgroundAudio != null) {
                _body.append("drop_background_audio", (_f = request.dropBackgroundAudio) === null || _f === void 0 ? void 0 : _f.toString());
            }
            if (request.useProfanityFilter != null) {
                _body.append("use_profanity_filter", (_g = request.useProfanityFilter) === null || _g === void 0 ? void 0 : _g.toString());
            }
            if (request.dubbingStudio != null) {
                _body.append("dubbing_studio", (_h = request.dubbingStudio) === null || _h === void 0 ? void 0 : _h.toString());
            }
            if (request.disableVoiceCloning != null) {
                _body.append("disable_voice_cloning", (_j = request.disableVoiceCloning) === null || _j === void 0 ? void 0 : _j.toString());
            }
            if (request.mode != null) {
                _body.append("mode", serializers.DubbingCreateRequestMode.jsonOrThrow(request.mode, { unrecognizedObjectKeys: "strip" }));
            }
            if (request.csvFps != null) {
                _body.append("csv_fps", (_k = request.csvFps) === null || _k === void 0 ? void 0 : _k.toString());
            }
            const _maybeEncodedRequest = yield _body.getRequest();
            const _headers = (0, headers_1.mergeHeaders)((_l = this._options) === null || _l === void 0 ? void 0 : _l.headers, (0, headers_1.mergeOnlyDefinedHeaders)(Object.assign({ "xi-api-key": (_m = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiKey) !== null && _m !== void 0 ? _m : (_o = this._options) === null || _o === void 0 ? void 0 : _o.apiKey }, _maybeEncodedRequest.headers)), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_p = this._options.fetcher) !== null && _p !== void 0 ? _p : core.fetcher)({
                url: core.url.join((_r = (_q = (yield core.Supplier.get(this._options.baseUrl))) !== null && _q !== void 0 ? _q : (yield core.Supplier.get(this._options.environment))) !== null && _r !== void 0 ? _r : environments.ElevenLabsEnvironment.Production, "v1/dubbing"),
                method: "POST",
                headers: _headers,
                queryParameters: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams,
                requestType: "file",
                duplex: _maybeEncodedRequest.duplex,
                body: _maybeEncodedRequest.body,
                timeoutMs: ((_u = (_s = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _s !== void 0 ? _s : (_t = this._options) === null || _t === void 0 ? void 0 : _t.timeoutInSeconds) !== null && _u !== void 0 ? _u : 240) * 1000,
                maxRetries: (_v = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _v !== void 0 ? _v : (_w = this._options) === null || _w === void 0 ? void 0 : _w.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_x = this._options) === null || _x === void 0 ? void 0 : _x.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.DoDubbingResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 422:
                        throw new ElevenLabs.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.ElevenLabsError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v1/dubbing");
        });
    }
    /**
     * Returns metadata about a dubbing project, including whether it's still in progress or not
     *
     * @param {string} dubbing_id - ID of the dubbing project.
     * @param {DubbingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link ElevenLabs.UnprocessableEntityError}
     *
     * @example
     *     await client.dubbing.get("dubbing_id")
     */
    get(dubbing_id, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__get(dubbing_id, requestOptions));
    }
    __get(dubbing_id, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
            const _headers = (0, headers_1.mergeHeaders)((_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "xi-api-key": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiKey) !== null && _b !== void 0 ? _b : (_c = this._options) === null || _c === void 0 ? void 0 : _c.apiKey }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_d = this._options.fetcher) !== null && _d !== void 0 ? _d : core.fetcher)({
                url: core.url.join((_f = (_e = (yield core.Supplier.get(this._options.baseUrl))) !== null && _e !== void 0 ? _e : (yield core.Supplier.get(this._options.environment))) !== null && _f !== void 0 ? _f : environments.ElevenLabsEnvironment.Production, `v1/dubbing/${core.url.encodePathParam(dubbing_id)}`),
                method: "GET",
                headers: _headers,
                queryParameters: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams,
                timeoutMs: ((_j = (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.timeoutInSeconds) !== null && _j !== void 0 ? _j : 240) * 1000,
                maxRetries: (_k = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _k !== void 0 ? _k : (_l = this._options) === null || _l === void 0 ? void 0 : _l.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_m = this._options) === null || _m === void 0 ? void 0 : _m.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.DubbingMetadataResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 422:
                        throw new ElevenLabs.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.ElevenLabsError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "GET", "/v1/dubbing/{dubbing_id}");
        });
    }
    /**
     * Deletes a dubbing project.
     *
     * @param {string} dubbing_id - ID of the dubbing project.
     * @param {DubbingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link ElevenLabs.UnprocessableEntityError}
     *
     * @example
     *     await client.dubbing.delete("dubbing_id")
     */
    delete(dubbing_id, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__delete(dubbing_id, requestOptions));
    }
    __delete(dubbing_id, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
            const _headers = (0, headers_1.mergeHeaders)((_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "xi-api-key": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiKey) !== null && _b !== void 0 ? _b : (_c = this._options) === null || _c === void 0 ? void 0 : _c.apiKey }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_d = this._options.fetcher) !== null && _d !== void 0 ? _d : core.fetcher)({
                url: core.url.join((_f = (_e = (yield core.Supplier.get(this._options.baseUrl))) !== null && _e !== void 0 ? _e : (yield core.Supplier.get(this._options.environment))) !== null && _f !== void 0 ? _f : environments.ElevenLabsEnvironment.Production, `v1/dubbing/${core.url.encodePathParam(dubbing_id)}`),
                method: "DELETE",
                headers: _headers,
                queryParameters: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams,
                timeoutMs: ((_j = (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.timeoutInSeconds) !== null && _j !== void 0 ? _j : 240) * 1000,
                maxRetries: (_k = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _k !== void 0 ? _k : (_l = this._options) === null || _l === void 0 ? void 0 : _l.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_m = this._options) === null || _m === void 0 ? void 0 : _m.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.DeleteDubbingResponseModel.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 422:
                        throw new ElevenLabs.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.ElevenLabsError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "DELETE", "/v1/dubbing/{dubbing_id}");
        });
    }
}
exports.DubbingClient = DubbingClient;
