// @elevenlabs/.ignored_elevenlabs-js/dist/serialization/types/AgentPlatformSettingsResponseModel.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentPlatformSettingsResponseModel = void 0;
const core = __importStar(require("../../core"));
const AgentCallLimits_1 = require("./AgentCallLimits");
const AgentTestingSettings_1 = require("./AgentTestingSettings");
const AgentTrustContext_1 = require("./AgentTrustContext");
const AgentWorkspaceOverridesOutput_1 = require("./AgentWorkspaceOverridesOutput");
const AnalysisProperty_1 = require("./AnalysisProperty");
const AnalysisScope_1 = require("./AnalysisScope");
const AuthSettings_1 = require("./AuthSettings");
const ConversationInitiationClientDataConfigOutput_1 = require("./ConversationInitiationClientDataConfigOutput");
const EvaluationSettingsOutput_1 = require("./EvaluationSettingsOutput");
const GuardrailsV1Output_1 = require("./GuardrailsV1Output");
const Llm_1 = require("./Llm");
const PrivacyConfigOutput_1 = require("./PrivacyConfigOutput");
const SafetyResponseModel_1 = require("./SafetyResponseModel");
const TopicDiscoverySettings_1 = require("./TopicDiscoverySettings");
const WidgetConfig_1 = require("./WidgetConfig");
exports.AgentPlatformSettingsResponseModel = core.serialization.object({
    evaluation: EvaluationSettingsOutput_1.EvaluationSettingsOutput.optional(),
    widget: WidgetConfig_1.WidgetConfig.optional(),
    dataCollection: core.serialization.property("data_collection", core.serialization.record(core.serialization.string(), AnalysisProperty_1.AnalysisProperty).optional()),
    dataCollectionScopes: core.serialization.property("data_collection_scopes", core.serialization.record(core.serialization.string(), AnalysisScope_1.AnalysisScope).optional()),
    overrides: ConversationInitiationClientDataConfigOutput_1.ConversationInitiationClientDataConfigOutput.optional(),
    workspaceOverrides: core.serialization.property("workspace_overrides", AgentWorkspaceOverridesOutput_1.AgentWorkspaceOverridesOutput.optional()),
    testing: AgentTestingSettings_1.AgentTestingSettings.optional(),
    archived: core.serialization.boolean().optional(),
    guardrails: GuardrailsV1Output_1.GuardrailsV1Output.optional(),
    summaryLanguage: core.serialization.property("summary_language", core.serialization.string().optional()),
    auth: AuthSettings_1.AuthSettings.optional(),
    callLimits: core.serialization.property("call_limits", AgentCallLimits_1.AgentCallLimits.optional()),
    privacy: PrivacyConfigOutput_1.PrivacyConfigOutput.optional(),
    trustContext: core.serialization.property("trust_context", AgentTrustContext_1.AgentTrustContext.optional()),
    analysisLlm: core.serialization.property("analysis_llm", Llm_1.Llm.optional()),
    topicDiscovery: core.serialization.property("topic_discovery", TopicDiscoverySettings_1.TopicDiscoverySettings.optional()),
    safety: SafetyResponseModel_1.SafetyResponseModel.optional(),
});
