// @elevenlabs/.ignored_elevenlabs-js/dist/serialization/types/ListAuthConnectionsResponseAuthConnectionsItem.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListAuthConnectionsResponseAuthConnectionsItem = void 0;
const core = __importStar(require("../../core"));
const ApiIntegrationOAuth2AuthCodeResponse_1 = require("./ApiIntegrationOAuth2AuthCodeResponse");
const ApiIntegrationOAuth2CustomAppResponse_1 = require("./ApiIntegrationOAuth2CustomAppResponse");
const BasicAuthResponse_1 = require("./BasicAuthResponse");
const BearerAuthResponse_1 = require("./BearerAuthResponse");
const CustomHeaderAuthResponse_1 = require("./CustomHeaderAuthResponse");
const MtlsAuthResponse_1 = require("./MtlsAuthResponse");
const OAuth2ClientCredsResponse_1 = require("./OAuth2ClientCredsResponse");
const OAuth2JwtResponse_1 = require("./OAuth2JwtResponse");
const PrivateKeyJwtResponse_1 = require("./PrivateKeyJwtResponse");
const SlackBotAuthResponse_1 = require("./SlackBotAuthResponse");
const UrlSecretAuthResponse_1 = require("./UrlSecretAuthResponse");
const WhatsAppAuthResponse_1 = require("./WhatsAppAuthResponse");
exports.ListAuthConnectionsResponseAuthConnectionsItem = core.serialization
    .union(core.serialization.discriminant("authType", "auth_type"), {
    api_integration_oauth2_auth_code: ApiIntegrationOAuth2AuthCodeResponse_1.ApiIntegrationOAuth2AuthCodeResponse,
    api_integration_oauth2_custom_app: ApiIntegrationOAuth2CustomAppResponse_1.ApiIntegrationOAuth2CustomAppResponse,
    basic_auth: BasicAuthResponse_1.BasicAuthResponse,
    bearer_auth: BearerAuthResponse_1.BearerAuthResponse,
    custom_header_auth: CustomHeaderAuthResponse_1.CustomHeaderAuthResponse,
    mtls: MtlsAuthResponse_1.MtlsAuthResponse,
    oauth2_client_credentials: OAuth2ClientCredsResponse_1.OAuth2ClientCredsResponse,
    oauth2_jwt: OAuth2JwtResponse_1.OAuth2JwtResponse,
    private_key_jwt: PrivateKeyJwtResponse_1.PrivateKeyJwtResponse,
    slack_bot_auth: SlackBotAuthResponse_1.SlackBotAuthResponse,
    url_secret: UrlSecretAuthResponse_1.UrlSecretAuthResponse,
    whatsapp_auth: WhatsAppAuthResponse_1.WhatsAppAuthResponse,
})
    .transform({
    transform: (value) => value,
    untransform: (value) => value,
});
