// @elevenlabs/.ignored_elevenlabs-js/dist/serialization/types/WorkspaceAuditLogEntryResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceAuditLogEntryResponse = void 0;
const core = __importStar(require("../../core"));
const ActorModel_1 = require("./ActorModel");
const DeviceModel_1 = require("./DeviceModel");
const HttpRequestModel_1 = require("./HttpRequestModel");
const SeverityId_1 = require("./SeverityId");
const StatusId_1 = require("./StatusId");
const WorkspaceAuditLogEntryResponseActivityId_1 = require("./WorkspaceAuditLogEntryResponseActivityId");
exports.WorkspaceAuditLogEntryResponse = core.serialization.object({
    metadata: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    time: core.serialization.number().optional(),
    activityId: core.serialization.property("activity_id", WorkspaceAuditLogEntryResponseActivityId_1.WorkspaceAuditLogEntryResponseActivityId),
    activityName: core.serialization.property("activity_name", core.serialization.string()),
    categoryName: core.serialization.property("category_name", core.serialization.string().optional()),
    categoryUid: core.serialization.property("category_uid", core.serialization.number().optional()),
    className: core.serialization.property("class_name", core.serialization.string().optional()),
    classUid: core.serialization.property("class_uid", core.serialization.number().optional()),
    severityId: core.serialization.property("severity_id", SeverityId_1.SeverityId.optional()),
    statusId: core.serialization.property("status_id", StatusId_1.StatusId),
    actor: ActorModel_1.ActorModel,
    device: DeviceModel_1.DeviceModel.optional(),
    httpRequest: core.serialization.property("http_request", HttpRequestModel_1.HttpRequestModel.optional()),
    message: core.serialization.string(),
    unmapped: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    id: core.serialization.string(),
    timeDt: core.serialization.property("time_dt", core.serialization.string().optional()),
    typeUid: core.serialization.property("type_uid", core.serialization.number().optional()),
    typeName: core.serialization.property("type_name", core.serialization.string().optional()),
});
