// @elevenlabs/.ignored_elevenlabs-js/serialization/types/CueOptionsRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CueOptionsRequest = void 0;
const core = __importStar(require("../../core"));
exports.CueOptionsRequest = core.serialization.object({
    minDurationMs: core.serialization.property("min_duration_ms", core.serialization.number().optional()),
    maxDurationMs: core.serialization.property("max_duration_ms", core.serialization.number().optional()),
    maxLinesPerCue: core.serialization.property("max_lines_per_cue", core.serialization.number().optional()),
    maxCharsPerLine: core.serialization.property("max_chars_per_line", core.serialization.number().optional()),
    maxCharsPerS: core.serialization.property("max_chars_per_s", core.serialization.number().optional()),
    minGapBetweenCuesFrames: core.serialization.property("min_gap_between_cues_frames", core.serialization.number().optional()),
});
