// @elevenlabs/.ignored_elevenlabs-js/serialization/types/LlmInfoModelOutput.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LlmInfoModelOutput = void 0;
const core = __importStar(require("../../core"));
const Llm_1 = require("./Llm");
const LlmDeprecationInfoModel_1 = require("./LlmDeprecationInfoModel");
const LlmReasoningEffort_1 = require("./LlmReasoningEffort");
const RegionalProcessingSurchargeInfo_1 = require("./RegionalProcessingSurchargeInfo");
exports.LlmInfoModelOutput = core.serialization.object({
    llm: Llm_1.Llm,
    isCheckpoint: core.serialization.property("is_checkpoint", core.serialization.boolean()),
    maxTokensLimit: core.serialization.property("max_tokens_limit", core.serialization.number()),
    maxContextLimit: core.serialization.property("max_context_limit", core.serialization.number()),
    supportsImageInput: core.serialization.property("supports_image_input", core.serialization.boolean()),
    supportsDocumentInput: core.serialization.property("supports_document_input", core.serialization.boolean()),
    supportsParallelToolCalls: core.serialization.property("supports_parallel_tool_calls", core.serialization.boolean()),
    availableReasoningEfforts: core.serialization.property("available_reasoning_efforts", core.serialization.list(LlmReasoningEffort_1.LlmReasoningEffort).optional()),
    deprecationInfo: core.serialization.property("deprecation_info", LlmDeprecationInfoModel_1.LlmDeprecationInfoModel.optional()),
    regionalProcessingSurcharge: core.serialization.property("regional_processing_surcharge", RegionalProcessingSurchargeInfo_1.RegionalProcessingSurchargeInfo.optional()),
});
