// @elevenlabs/.ignored_elevenlabs-js/serialization/types/OAuth2JwtResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuth2JwtResponse = void 0;
const core = __importStar(require("../../core"));
const AuthConnectionDependencies_1 = require("./AuthConnectionDependencies");
const AuthConnectionStatus_1 = require("./AuthConnectionStatus");
const OAuth2JwtResponseAlgorithm_1 = require("./OAuth2JwtResponseAlgorithm");
const OAuth2JwtResponseTokenResponseField_1 = require("./OAuth2JwtResponseTokenResponseField");
exports.OAuth2JwtResponse = core.serialization.object({
    name: core.serialization.string(),
    provider: core.serialization.string(),
    algorithm: OAuth2JwtResponseAlgorithm_1.OAuth2JwtResponseAlgorithm.optional(),
    keyId: core.serialization.property("key_id", core.serialization.string().optional()),
    issuer: core.serialization.string(),
    audience: core.serialization.string(),
    subject: core.serialization.string(),
    expirationSeconds: core.serialization.property("expiration_seconds", core.serialization.number().optional()),
    extraParams: core.serialization.property("extra_params", core.serialization.record(core.serialization.string(), core.serialization.string()).optional()),
    tokenUrl: core.serialization.property("token_url", core.serialization.string()),
    scopes: core.serialization.list(core.serialization.string()).optional(),
    tokenResponseField: core.serialization.property("token_response_field", OAuth2JwtResponseTokenResponseField_1.OAuth2JwtResponseTokenResponseField.optional()),
    id: core.serialization.string(),
    usedBy: core.serialization.property("used_by", AuthConnectionDependencies_1.AuthConnectionDependencies.optional()),
    status: AuthConnectionStatus_1.AuthConnectionStatus.optional(),
    statusDetail: core.serialization.property("status_detail", core.serialization.string().optional()),
    statusUpdatedAt: core.serialization.property("status_updated_at", core.serialization.string().optional()),
});
