// @elevenlabs/.ignored_elevenlabs-js/serialization/types/ProjectVideoResponseModel.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectVideoResponseModel = void 0;
const core = __importStar(require("../../core"));
const AssetTranscription_1 = require("./AssetTranscription");
const CanvasPlacement_1 = require("./CanvasPlacement");
const ClipAnimation_1 = require("./ClipAnimation");
const GenerationSourceContext_1 = require("./GenerationSourceContext");
const PendingBlocksMetadataModel_1 = require("./PendingBlocksMetadataModel");
const PendingClipTask_1 = require("./PendingClipTask");
const PendingExternalAudiosMetadataModel_1 = require("./PendingExternalAudiosMetadataModel");
const ProjectVideoThumbnailSheetResponseModel_1 = require("./ProjectVideoThumbnailSheetResponseModel");
const VideoAnalysis_1 = require("./VideoAnalysis");
exports.ProjectVideoResponseModel = core.serialization.object({
    videoId: core.serialization.property("video_id", core.serialization.string()),
    filename: core.serialization.string(),
    signedUrl: core.serialization.property("signed_url", core.serialization.string().optional()),
    signedPreviewUrl: core.serialization.property("signed_preview_url", core.serialization.string().optional()),
    offsetMs: core.serialization.property("offset_ms", core.serialization.number()),
    durationMs: core.serialization.property("duration_ms", core.serialization.number()),
    volumeGainDb: core.serialization.property("volume_gain_db", core.serialization.number()),
    muted: core.serialization.boolean(),
    fadeInMs: core.serialization.property("fade_in_ms", core.serialization.number().optional()),
    fadeOutMs: core.serialization.property("fade_out_ms", core.serialization.number().optional()),
    width: core.serialization.number(),
    height: core.serialization.number(),
    codec: core.serialization.string(),
    order: core.serialization.string(),
    createdAtMs: core.serialization.property("created_at_ms", core.serialization.number()),
    updatedAtMs: core.serialization.property("updated_at_ms", core.serialization.number()),
    error: core.serialization.string().optional(),
    thumbnailIntervalSeconds: core.serialization.property("thumbnail_interval_seconds", core.serialization.number()),
    thumbnailSize: core.serialization.property("thumbnail_size", core.serialization.list(core.serialization.number())),
    thumbnailSheets: core.serialization.property("thumbnail_sheets", core.serialization.list(ProjectVideoThumbnailSheetResponseModel_1.ProjectVideoThumbnailSheetResponseModel)),
    startTimeMs: core.serialization.property("start_time_ms", core.serialization.number()),
    endTimeMs: core.serialization.property("end_time_ms", core.serialization.number().optional()),
    assetPreviewSignedUrl: core.serialization.property("asset_preview_signed_url", core.serialization.string().optional()),
    sourceVideoId: core.serialization.property("source_video_id", core.serialization.string().optional()),
    sourceAssetId: core.serialization.property("source_asset_id", core.serialization.string().optional()),
    sourcePlatformAssetId: core.serialization.property("source_platform_asset_id", core.serialization.string().optional()),
    pendingBlocksMetadata: core.serialization.property("pending_blocks_metadata", PendingBlocksMetadataModel_1.PendingBlocksMetadataModel.optional()),
    pendingExternalAudiosMetadata: core.serialization.property("pending_external_audios_metadata", PendingExternalAudiosMetadataModel_1.PendingExternalAudiosMetadataModel.optional()),
    speechImported: core.serialization.property("speech_imported", core.serialization.boolean().optional()),
    pendingTask: core.serialization.property("pending_task", PendingClipTask_1.PendingClipTask.optional()),
    audioTrackReady: core.serialization.property("audio_track_ready", core.serialization.boolean().optional()),
    exportFormatReady: core.serialization.property("export_format_ready", core.serialization.boolean().optional()),
    currentSnapshotId: core.serialization.property("current_snapshot_id", core.serialization.string().optional()),
    sourceContext: core.serialization.property("source_context", GenerationSourceContext_1.GenerationSourceContext.optional()),
    analysis: VideoAnalysis_1.VideoAnalysis.optional(),
    transcription: AssetTranscription_1.AssetTranscription.optional(),
    canvasPlacement: core.serialization.property("canvas_placement", CanvasPlacement_1.CanvasPlacement.optional()),
    animation: ClipAnimation_1.ClipAnimation.optional(),
    playbackSpeed: core.serialization.property("playback_speed", core.serialization.number().optional()),
    opacity: core.serialization.number().optional(),
    trackId: core.serialization.property("track_id", core.serialization.string().optional()),
    previewJobProgress: core.serialization.property("preview_job_progress", core.serialization.number().optional()),
    importSpeechProgress: core.serialization.property("import_speech_progress", core.serialization.number().optional()),
});
