// @elevenlabs/.ignored_elevenlabs-js/serialization/types/SpeechEngineResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeechEngineResponse = void 0;
const core = __importStar(require("../../core"));
const AgentCallLimits_1 = require("./AgentCallLimits");
const AgentMetadataDbModel_1 = require("./AgentMetadataDbModel");
const AsrConversationalConfig_1 = require("./AsrConversationalConfig");
const BaseTurnConfig_1 = require("./BaseTurnConfig");
const ConversationConfigOutput_1 = require("./ConversationConfigOutput");
const PrivacyConfigOutput_1 = require("./PrivacyConfigOutput");
const ResourceAccessInfo_1 = require("./ResourceAccessInfo");
const SpeechEngineConfig_1 = require("./SpeechEngineConfig");
const SpeechEngineConversationInitiationClientDataConfig_1 = require("./SpeechEngineConversationInitiationClientDataConfig");
const TtsConversationalConfigOutput_1 = require("./TtsConversationalConfigOutput");
exports.SpeechEngineResponse = core.serialization.object({
    speechEngineId: core.serialization.property("speech_engine_id", core.serialization.string()),
    name: core.serialization.string(),
    speechEngine: core.serialization.property("speech_engine", SpeechEngineConfig_1.SpeechEngineConfig),
    asr: AsrConversationalConfig_1.AsrConversationalConfig,
    tts: TtsConversationalConfigOutput_1.TtsConversationalConfigOutput,
    turn: BaseTurnConfig_1.BaseTurnConfig,
    conversation: ConversationConfigOutput_1.ConversationConfigOutput,
    privacy: PrivacyConfigOutput_1.PrivacyConfigOutput,
    callLimits: core.serialization.property("call_limits", AgentCallLimits_1.AgentCallLimits),
    language: core.serialization.string(),
    tags: core.serialization.list(core.serialization.string()),
    overrides: SpeechEngineConversationInitiationClientDataConfig_1.SpeechEngineConversationInitiationClientDataConfig,
    metadata: AgentMetadataDbModel_1.AgentMetadataDbModel,
    accessInfo: core.serialization.property("access_info", ResourceAccessInfo_1.ResourceAccessInfo.optional()),
});
