// @elevenlabs/.ignored_elevenlabs-js/serialization/types/TurnConfig.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurnConfig = void 0;
const core = __importStar(require("../../core"));
const SoftTimeoutConfig_1 = require("./SoftTimeoutConfig");
const SpellingPatience_1 = require("./SpellingPatience");
const TurnEagerness_1 = require("./TurnEagerness");
const TurnModel_1 = require("./TurnModel");
exports.TurnConfig = core.serialization.object({
    turnTimeout: core.serialization.property("turn_timeout", core.serialization.number().optional()),
    initialWaitTime: core.serialization.property("initial_wait_time", core.serialization.number().optional()),
    silenceEndCallTimeout: core.serialization.property("silence_end_call_timeout", core.serialization.number().optional()),
    turnEagerness: core.serialization.property("turn_eagerness", TurnEagerness_1.TurnEagerness.optional()),
    spellingPatience: core.serialization.property("spelling_patience", SpellingPatience_1.SpellingPatience.optional()),
    speculativeTurn: core.serialization.property("speculative_turn", core.serialization.boolean().optional()),
    retranscribeOnTurnTimeout: core.serialization.property("retranscribe_on_turn_timeout", core.serialization.boolean().optional()),
    turnModel: core.serialization.property("turn_model", TurnModel_1.TurnModel.optional()),
    interruptionIgnoreTerms: core.serialization.property("interruption_ignore_terms", core.serialization.list(core.serialization.string()).optional()),
    transcribeOnDisabledInterruptions: core.serialization.property("transcribe_on_disabled_interruptions", core.serialization.boolean().optional()),
    softTimeoutConfig: core.serialization.property("soft_timeout_config", SoftTimeoutConfig_1.SoftTimeoutConfig.optional()),
});
